core/src/main/scala/org/apache/spark/util/random/SamplingUtils.scala

```diff
@@ -52,16 +52,21 @@ private[spark] object SamplingUtils {
       (trimReservoir, i)
     } else {
       // If input size > k, continue the sampling process.
+      var l = i.toLong
       val rand = new XORShiftRandom(seed)
       while (input.hasNext) {
         val item = input.next()
-        val replacementIndex = rand.nextInt(i)
+        val replacementIndex = if (l < Int.MaxValue) {
+          rand.nextInt(l.toInt)
+        } else {
+          rand.nextLong()
```

Comment:
This isn't valid, because it chooses a value from (negative) `Long.MIN_VALUE` to `Long.MAX_VALUE`. You need to choose a number in `[0, l)`.

Rather than even special case it, just use

```
val replacementIndex = (random.nextDouble() * l).toLong
```